fluss-common/src/main/java/com/alibaba/fluss/memory/LazyMemorySegmentPool.java

```diff
@@ -107,38 +109,33 @@ public static LazyMemorySegmentPool create(Configuration conf) {
 
     @Override
     public MemorySegment nextSegment(boolean waiting) {
-        return inLock(
-                lock,
-                () -> {
-                    checkClosed();
-                    int freePages = freePages();
-                    if (freePages == 0) {
-                        if (waiting) {
-                            return waitForSegment();
-                        } else {
-                            return null;
-                        }
-                    }
-
-                    if (cachePages.isEmpty()) {
-                        int numPages = Math.min(freePages, perRequestPages);
-                        for (int i = 0; i < numPages; i++) {
-                            cachePages.add(MemorySegment.allocateHeapMemory(pageSize));
-                        }
-                    }
-
-                    this.pageUsage++;
-                    return cachePages.remove(this.cachePages.size() - 1);
-                });
+        checkClosed();
+        int freePages = freePages();
+        if (freePages == 0) {
+            if (waiting) {
+                return waitForSegment();
+            } else {
+                return null;
+            }
+        }
+
+        if (cachePages.isEmpty()) {
+            int numPages = Math.min(freePages, perRequestPages);
+            for (int i = 0; i < numPages; i++) {
+                cachePages.add(MemorySegment.allocateHeapMemory(pageSize));
+            }
+        }
+
+        this.pageUsage++;
```

Comment:
if no lock here, maybe two theads try to get nextSegment. Then all freePages return is empty, then allocateHeapMemory twice, maybe out of bound. By the way `this.pageUsage++` is also not thread safe operation.

Reply:
Yes, I noticed that, and I have submitted a new version of the code